liblwgeom/lwgeom_transform.c

```diff
@@ -209,202 +150,94 @@ projpj_from_string(const char *str1)
-	double out_semi_major_metre = DBL_MAX, out_semi_minor_metre = DBL_MAX;
+	double semi_major_metre = DBL_MAX, semi_minor_metre = DBL_MAX;
 
-	if (!pj_source_crs)
-	{
-		lwerror("%s: unable to access source crs", __func__);
+	/* Usable inputs? */
+	if (! (str_in && str_out))
 		return NULL;
-	}
-	uint8_t source_swapped = proj_crs_is_swapped(pj_source_crs);
 
-	/* We only care about the extra values if there is no transformation */
-	if (!extra_geography_data)
-	{
-		proj_destroy(pj_source_crs);
-	}
-	else
+	PJ* pj = proj_create_crs_to_crs(PJ_DEFAULT_CTX, str_in, str_out, NULL);
+	if (!pj)
+		return NULL;
+
+	/* Fill in geodetic parameter information when a null-transform */
+	/* is passed, because that's how we signal we want to store */
+	/* that info in the cache */
+	if (strcmp(str_in, str_out) == 0)
 	{
+		PJ *pj_source_crs = proj_get_source_crs(PJ_DEFAULT_CTX, pj);
 		PJ *pj_ellps;
-		double out_inv_flattening;
-		int out_is_semi_minor_computed;
-
 		PJ_TYPE pj_type = proj_get_type(pj_source_crs);
 		if (pj_type == PJ_TYPE_UNKNOWN)
 		{
-			proj_destroy(pj_source_crs);
+			proj_destroy(pj);
 			lwerror("%s: unable to access source crs type", __func__);
 			return NULL;
 		}
 		source_is_latlong = (pj_type == PJ_TYPE_GEOGRAPHIC_2D_CRS) || (pj_type == PJ_TYPE_GEOGRAPHIC_3D_CRS);
 
-		pj_ellps = proj_get_ellipsoid(NULL, pj_source_crs);
+		pj_ellps = proj_get_ellipsoid(PJ_DEFAULT_CTX, pj_source_crs);
 		proj_destroy(pj_source_crs);
 		if (!pj_ellps)
 		{
+			proj_destroy(pj);
 			lwerror("%s: unable to access source crs ellipsoid", __func__);
 			return NULL;
 		}
-		if (!proj_ellipsoid_get_parameters(NULL,
+		if (!proj_ellipsoid_get_parameters(PJ_DEFAULT_CTX,
 						   pj_ellps,
-						   &out_semi_major_metre,
-						   &out_semi_minor_metre,
-						   &out_is_semi_minor_computed,
-						   &out_inv_flattening))
+						   &semi_major_metre,
+						   &semi_minor_metre,
+						   NULL,
+						   NULL))
 		{
 			proj_destroy(pj_ellps);
+			proj_destroy(pj);
 			lwerror("%s: unable to access source crs ellipsoid parameters", __func__);
 			return NULL;
 		}
 		proj_destroy(pj_ellps);
 	}
 
-	PJ *pj_target_crs = proj_get_target_crs(NULL, pj);
-	if (!pj_target_crs)
-	{
-		lwerror("%s: unable to access target crs", __func__);
-		return NULL;
-	}
-	uint8_t target_swapped = proj_crs_is_swapped(pj_target_crs);
-	proj_destroy(pj_target_crs);
+	/* Add in an axis swap if necessary */
+	PJ* pj_norm = proj_normalize_for_visualization(PJ_DEFAULT_CTX, pj);
+	/* Swap failed for some reason? Fall back to coordinate operation */
+	if (!pj_norm)
+		pj_norm = pj;
+	/* Swap is not a copy of input? Clean up input */
+	else if (pj != pj_norm)
+		proj_destroy(pj);
 
+	/* Allocate and populate return value */
 	LWPROJ *lp = lwalloc(sizeof(LWPROJ));
-	lp->pj = pj;
-	lp->source_swapped = source_swapped;
-	lp->target_swapped = target_swapped;
+	lp->pj = pj_norm; /* Caller is going to have to explicitly proj_destroy this */
+	lp->source_is_latlong = LW_FALSE;
```

Comment:
This looks odd, 2 consecutive assigments

Reply:
Yep, little copy pasta, thanks.